datalad/cmd.py

```diff
@@ -98,6 +98,221 @@ def _cleanup_output(stream, std):
+          provided as stdout return value.
+          The helper functions 'kill_output' and 'capture_output' are
+          provided to either swallow all output (and not relay it to the
+          parent) or to capture all output and provide it as the return
+          value.
+        proc_stderr : callable, optional
+          Like proc_stdout, but for stderr.
+        stdin : byte stream, optional
+          File descriptor like, used as stdin for the process. Passed
+          verbatim to subprocess.Popen().
+        poll_latency : float, optional
+          Shortest interval at which the running process is queried for
+          output (in seconds). Any potential output processing will increase
+          the effective interval. When the effective polling frequency is
+          too low to keep the output buffers below their maximum size,
+          a process will deadlock.
+
+        Returns
+        -------
+        stdout, stderr
+          Unicode string with the cumulative standard output and error
+          of the process.
+
+        Raises
+        ------
+        CommandError
+          On execution failure (non-zero exit code) this exception is
+          raised which provides the command (cmd), stdout, stderr,
+          exit code (status), and a message identifying the failed
+          command, as properties.
+        """
+        proc_out = (proc_stdout, proc_stderr)
+        try:
+            lgr.log(8, "Start running %r", cmd)
+            process = subprocess.Popen(
+                cmd,
+                stdout=subprocess.PIPE if proc_stdout else None,
+                stderr=subprocess.PIPE if proc_stderr else None,
+                shell=False,
+                cwd=self.cwd,
+                env=self.env,
+                stdin=stdin,
+                # intermediate reports are never decoded anyways
+                # from PY37 onwards
+                #text=False,
+                universal_newlines=False,
+            )
+        except Exception as e:
+            lgr.log(11, "Failed to start %r%r: %s" %
+                    (cmd,
+                     (" under %r" % self.cwd) if self.cwd else '',
+                     exc_str(e)))
+            raise
+
+        try:
+            out = [[], []]
+            unprocessed = [None, None]
+            # make sure to run this loop at least once, even if the
+            # process is already dead
+            keep_going = True
+
+            def _handle_output(u, o, p):
+                processed, unprocessed_len = proc(u)
+                if processed:
+                    o.append(processed)
+                return u[-(unprocessed_len):] if unprocessed_len else None
+
+            def _read_stream(stream):
+                #if sys.version_info.major == 3 and sys.version_info.minor < 7:
+                nbytes_avail = len(stream.peek())
+                return stream.read(nbytes_avail) if nbytes_avail > 0 else None
+                # o.read1(-1)
```

Comment:
It looks like the negative `size` support added in Python's ccb2c0e310 (Implement optional BufferedReader, BytesIO read1() argument, 2016-10-20) is straightforward:

```diff
diff --git a/Lib/_pyio.py b/Lib/_pyio.py
index d0947f06d5..569527bc37 100644
--- a/Lib/_pyio.py
+++ b/Lib/_pyio.py
@@ -1073,12 +1073,12 @@ def _peek_unlocked(self, n=0):
                 self._read_pos = 0
         return self._read_buf[self._read_pos:]
 
-    def read1(self, size):
+    def read1(self, size=-1):
         """Reads up to size bytes, with at most one read() system call."""
         # Returns up to size bytes.  If at least one byte is buffered, we
         # only return buffered bytes.  Otherwise, we do one raw read.
         if size < 0:
-            raise ValueError("number of bytes to read must be positive")
+            size = self.buffer_size
         if size == 0:
             return b""
         with self._read_lock:
```

So, to get the same behavior before 3.7, can't you use `.buffer_size` here rather than the `peek()` call?

Reply:
I have adjusted the comment. I came to realize that `read(-1)` will still make a blocking read (when the buffer happens to be empty), which can lead to process deadlock, when the other output channel is flooded with data in the meantime.